pkg/azure/aks/pricing_map.go

```diff
@@ -0,0 +1,187 @@
+)
+
+type MachineOperatingSystem int
+
+const (
+	Linux MachineOperatingSystem = iota
+	Windows
+)
+
+var machineOperatingSystemNames [2]string = [2]string{"Linux", "Windows"}
+
+func (o MachineOperatingSystem) String() string {
+	return machineOperatingSystemNames[o-1]
+}
+
+type MachinePriority int
+
+const (
+	OnDemand MachinePriority = iota
+	Spot
+)
+
+var machinePriorityNames [2]string = [2]string{"OnDemand", "Spot"}
+
+func (v MachinePriority) String() string {
+	return machinePriorityNames[v-1]
+}
+
+type PriceBySku map[string]retailPriceSdk.ResourceSKU
+
+type PriceByOperatingSystem map[MachineOperatingSystem]PriceBySku
+
+type PriceByPriority map[MachinePriority]PriceByOperatingSystem
+
+type PriceStore struct {
+	lock              *sync.RWMutex
+	subscriptionId    string
+	logger            *slog.Logger
+	context           context.Context
+	retailPriceClient *retailPriceSdk.RetailPricesClient
+
+	RegionMap map[string]PriceByPriority
+	Cache     map[string]*retailPriceSdk.ResourceSKU
+}
+
+func NewPricingStore(subId string, priceClient *retailPriceSdk.RetailPricesClient, parentLogger *slog.Logger, parentContext context.Context) *PriceStore {
+	logger := parentLogger.With("subsystem", "pricingMap")
+
+	p := &PriceStore{
+		lock:              &sync.RWMutex{},
+		logger:            logger,
+		context:           parentContext,
+		subscriptionId:    subId,
+		retailPriceClient: priceClient,
+
+		RegionMap: make(map[string]PriceByPriority),
+		Cache:     make(map[string]*retailPriceSdk.ResourceSKU),
+	}
+
+	go func() {
+		err := p.PopulatePriceStore([]string{})
+		if err != nil {
+			p.logger.LogAttrs(p.context, slog.LevelError, "error populating initial price store", slog.String("error", err.Error()))
+		}
+	}()
+
+	return p
+}
+
+func (p *PriceStore) buildQueryFilter(locationList []string) string {
+	if len(locationList) == 0 {
+		return `serviceName eq 'Virtual Machines' and priceType eq 'Consumption'`
+	}
+
+	locationListFilter := []string{}
+	for _, region := range locationList {
+		locationListFilter = append(locationListFilter, fmt.Sprintf("armRegionName eq '%s'", region))
+	}
+
+	locationListStr := strings.Join(locationListFilter, " or ")
+	return fmt.Sprintf(`serviceName eq 'Virtual Machines' and priceType eq 'Consumption' and (%s)`, locationListStr)
+}
+
+func (p *PriceStore) buildListOptions(locationList []string) *retailPriceSdk.RetailPricesClientListOptions {
+	queryFilter := p.buildQueryFilter(locationList)
+	return &retailPriceSdk.RetailPricesClientListOptions{
+		APIVersion:  to.StringPtr(AZ_API_VERSION),
+		Filter:      to.StringPtr(queryFilter),
+		MeterRegion: to.StringPtr(`'primary'`),
+	}
+}
+
+func (p *PriceStore) determineMachineOperatingSystem(sku retailPriceSdk.ResourceSKU) MachineOperatingSystem {
+	switch {
+	case strings.Contains(sku.ProductName, "Windows"):
+		return Windows
+	default:
+		return Linux
+	}
+}
+
+func (p *PriceStore) determineMachinePriority(sku retailPriceSdk.ResourceSKU) MachinePriority {
```

Comment:
After seeing the usage via tests, it's pretty clear this likely doesn't need to be a method of `PriceStore` and would be better off having it as a private method for the `aks` module.